server/executor/tracepollerworker/starter_worker.go

```diff
@@ -85,29 +100,52 @@ func (w *tracePollerStarterWorker) ProcessItem(ctx context.Context, job executor
 
 	emitEvent(ctx, w.state, events.TraceFetchingStart(job.Test.ID, job.Run.ID))
 
-	err = w.testConnection(ctx, traceDB, &job)
+	endpoints := traceDB.GetEndpoints()
+	ds, err := w.state.dsRepo.Current(ctx)
+	if err != nil {
+		wrappedError := fmt.Errorf("could not get current datastore: %w", err)
+		handleError(ctx, job, wrappedError, w.state, span)
+		return
+	}
+
+	connectionResult, err := w.testConnection(ctx, traceDB, ds)
 	if err != nil {
+		log.Printf("[TracePoller] TestConnection error: %s", err.Error())
 		handleError(ctx, job, err, w.state, span)
 		return
 	}
 
+	if connectionResult != nil {
+		emitEvent(ctx, w.state, events.TraceDataStoreConnectionInfo(job.Test.ID, job.Run.ID, *connectionResult))
+	}
+
+	emitEvent(ctx, w.state, events.TracePollingStart(job.Test.ID, job.Run.ID, string(ds.Type), endpoints))
+
 	w.outputQueue.Enqueue(ctx, job)
 }
 
-func (w *tracePollerStarterWorker) testConnection(ctx context.Context, traceDB tracedb.TraceDB, job *executor.Job) error {
-	if testableTraceDB, ok := traceDB.(tracedb.TestableTraceDB); ok {
-		connectionResult := testableTraceDB.TestConnection(ctx)
-
-		emitEvent(ctx, w.state, events.TraceDataStoreConnectionInfo(job.Test.ID, job.Run.ID, connectionResult))
+func (w *tracePollerStarterWorker) testConnection(ctx context.Context, traceDB tracedb.TraceDB, ds datastore.DataStore) (*model.ConnectionResult, error) {
+	_, ok := traceDB.(tracedb.TestableTraceDB)
+	if !ok {
+		return nil, nil
 	}
 
-	endpoints := traceDB.GetEndpoints()
-	ds, err := w.state.dsRepo.Current(ctx)
+	job := w.dsTestPipeline.NewJob(ctx, ds)
+
+	wg := sync.WaitGroup{}
+	err := w.dsTestPipeline.Subscribe(job.ID, func(result testconnection.Job) {
+		job = result
+		wg.Done()
+	})
+
 	if err != nil {
-		return fmt.Errorf("could not get current datastore: %w", err)
+		return nil, err
 	}
 
-	emitEvent(ctx, w.state, events.TracePollingStart(job.Test.ID, job.Run.ID, string(ds.Type), endpoints))
+	w.dsTestPipeline.Run(ctx, job)
+	wg.Add(1)
```

Comment:
```suggestion
   	wg.Add(1)
	w.dsTestPipeline.Run(ctx, job)
```

This is important because if you run `wg.Done()` before `wg.Add()` is executed, the app will panic. I know this should not happen in normal cases, but we never know. Better be safe than sorry.

https://go.dev/play/p/NHIkZ00KhGx